File: programming-paradigm/object-oriented-programming/polymorphism/AnimalExample.py
# -*- coding: utf-8 -*-
'''
# Python实现多态例子
'''

# 定义一个通用Animal类  

class Animal:  
    def make_sound(self):  
        print("Animal makes a sound")  
  
# 定义Dog类，继承Animal
class Dog(Animal):  
    name = "Dog"
    def make_sound(self):  
        print("Dog barks")  
  
# 定义Cat类，继承Animal
class Cat(Animal):  
    name = "Cat"
    def make_sound(self):  
        print("Cat meows")  
  
    # Cat自有方法  
    def meow(self):  
        print(self.name + " is meowing...")  
  
# 定义Bird类，它是动物的子类  
class Bird(Animal):  
    def make_sound(self):  
        print("Bird chirps")  

# 定义管理类
class Zoo:  
    # python与js一样为动态语言，使用duck typing，不需要显式声明接口
    def let_animal_make_sound(self, animal):  
        animal.make_sound()  
  
# 测试代码
if __name__ == "__main__":
    zoo = Zoo()

    # 直接创建实例，Python中不需要向上转型
    my_dog = Dog()
    my_cat = Cat()
    my_bird = Bird()

    # 直接调用自有方法
    my_cat.meow()

    # Python中可直接给对象设置方法，self不会改变
    my_dog.meow = my_cat.meow
    my_dog.meow()

    # 动物园可以使用相同的方法处理不同种类的动物
    zoo.let_animal_make_sound(my_dog)  # 输出 "Dog barks"
    zoo.let_animal_make_sound(my_cat)  # 输出 "Cat meows"
    zoo.let_animal_make_sound(my_bird)  # 输出 "Bird chirps"

"""
jarry@jarrys-MacBook-Pro polymorphism % python AnimalExample.py
Cat is meowing...
Cat is meowing...
Dog barks
Cat meows
Bird chirps
"""

File: programming-paradigm/object-oriented-programming/polymorphism/AnimalExample.java
/*Java实现多态的例子 */

// 定义一个通用Animal类
class Animal {
  void makeSound() {
    System.out.println("Animal makes a sound");
  }
}

// 或者定义Animal接口，子类来实现
// interface Animal {
// void makeSound();
// }

// 定义Dog类，它是动物的子类
class Dog extends Animal {
  @Override
  void makeSound() {
    System.out.println("Dog barks");
  }
}

// 定义Cat类，它是动物的子类
class Cat extends Animal {
  @Override
  void makeSound() {
    System.out.println("Cat meows");
  }

  // Cat自有方法
  void meow() {
    System.out.println("Cat is meowing...");
  }
}

// 定义一个动物园类，管理不同的动物
class Zoo {
  // 传入的是抽象父类或接口，方便扩展
  void letAnimalMakeSound(Animal animal) {
    animal.makeSound();
  }
}

class Bird extends Animal {
  @Override
  void makeSound() {
    System.out.println("Bird chirps");
  }
}

public class AnimalExample {
  public static void main(String[] args) {
    Zoo zoo = new Zoo();

    Animal myDog = new Dog(); // 向上转型
    Animal myCat = new Cat(); // 向上转型
    ((Cat)myCat).meow(); // 向下强转，打印自有方法
    Animal myBird = new Bird(); // 向上转型

    // 通过多态性，动物园可以使用相同的方法处理不同种类的动物
    zoo.letAnimalMakeSound(myDog); // 输出 "Dog barks"
    zoo.letAnimalMakeSound(myCat); // 输出 "Cat meows"
    zoo.letAnimalMakeSound(myBird); // 输出 "Bird chirps"
  }
}